pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java

```diff
@@ -0,0 +1,272 @@
+      if (operands.stream().anyMatch(operand -> operand.equals(TRUE))) {
+        return setExpressionToBoolean(expression, true);
+      }
+
+      // Remove all Literal operands that are FALSE.
+      operands.removeIf(x -> x.equals(FALSE));
+      if (operands.size() == 0) {
+        return setExpressionToBoolean(expression, false);
+      }
+    }
+
+    return expression;
+  }
+
+  private boolean isNumericColumn(Expression expression, Schema schema) {
+    if (expression.getType() != ExpressionType.IDENTIFIER) {
+      // Expression can not be a column.
+      return false;
+    }
+
+    String column = expression.getIdentifier().getName();
+    FieldSpec fieldSpec = schema.getFieldSpecFor(column);
+    if (fieldSpec == null || !fieldSpec.isSingleValueField()) {
+      // Expression can not be a column name.
+      return false;
+    }
+
+    return schema.getFieldSpecFor(column).getDataType().isNumeric();
+  }
+
+  private boolean isNumericLiteral(Expression expression) {
+    if (expression.getType() == ExpressionType.LITERAL) {
+      Literal._Fields type = expression.getLiteral().getSetField();
+      switch (type) {
+        case SHORT_VALUE:
+        case INT_VALUE:
+        case LONG_VALUE:
+        case DOUBLE_VALUE:
+          return true;
+      }
+    }
+    return false;
+  }
+
+  /** Change the expression value to boolean literal with given value. */
+  private static Expression setExpressionToBoolean(Expression expression, boolean value) {
+    expression.unsetFunctionCall();
+    expression.setType(ExpressionType.LITERAL);
+    expression.setLiteral(Literal.boolValue(value));
+
+    return expression;
+  }
+
+  /**
+   * Rewrite expressions of form "column = literal" or "column != literal" to ensure that RHS literal is the same
+   * datatype as LHS column.
+   */
+  private Expression rewrite(Expression equals, Expression lhs, Expression rhs, Schema schema) {
+    // Get expression operator
+    boolean result = equals.getFunctionCall().getOperator().equals(FilterKind.NOT_EQUALS.name());
+
+    // Get column data type.
+    FieldSpec.DataType dataType = schema.getFieldSpecFor(lhs.getIdentifier().getName()).getDataType();
+
+    switch (rhs.getLiteral().getSetField()) {
+      case SHORT_VALUE:
+      case INT_VALUE:
+        // No rewrites needed since SHORT and INT conversion to numeric column types (INT, LONG, FLOAT, and DOUBLE) is
+        // lossless and will be implicitly handled on the server side.
+        break;
+      case LONG_VALUE: {
+        long actual = rhs.getLiteral().getLongValue();
+        switch (dataType) {
+          case INT: {
+            int converted = (int) actual;
+            if (converted != actual) {
+              // Long value does not fall within the bounds of INT column.
+              setExpressionToBoolean(equals, result);
+            } else {
+              // Replace long value with converted int value.
+              rhs.setLiteral(Literal.intValue(converted));
```

Comment:
Use `RequestUtils.getLiteralExpression()` to create literal. Same for other places

Reply:
Did in place modification since we need to preserve the reference to `rhs`  to avoid destroying the expression tree that is being traversed.